src/hpa.c

```diff
@@ -0,0 +1,395 @@
+		assert(edata_state_get(edata) == extent_state_active);
+		assert(edata_arena_ind_get(edata) == shard->ind);
+		assert(edata_szind_get_maybe_invalid(edata) == SC_NSIZES);
+		assert(!edata_slab_get(edata));
+		assert(edata_committed_get(edata));
+		assert(edata_base_get(edata) == edata_addr_get(edata));
+		assert(edata_base_get(edata) != NULL);
+	}
+	return edata;
+}
+
+static bool
+hpa_expand(tsdn_t *tsdn, pai_t *self, edata_t *edata,
+    size_t old_size, size_t new_size, bool zero) {
+	/* Expand not yet supported. */
+	return true;
+}
+
+static bool
+hpa_shrink(tsdn_t *tsdn, pai_t *self, edata_t *edata,
+    size_t old_size, size_t new_size) {
+	/* Shrink not yet supported. */
+	return true;
+}
+
+static void
+hpa_dalloc_central(tsdn_t *tsdn, hpa_shard_t *shard, edata_t *edata) {
+	hpa_t *hpa = shard->hpa;
+
+	edata_arena_ind_set(edata, hpa->ind);
+	malloc_mutex_lock(tsdn, &hpa->mtx);
+	hpa_central_dalloc(tsdn, &hpa->central, edata);
+	malloc_mutex_unlock(tsdn, &hpa->mtx);
+}
+
+static void
+hpa_dalloc(tsdn_t *tsdn, pai_t *self, edata_t *edata) {
+	hpa_shard_t *shard = (hpa_shard_t *)self;
+
+	edata_addr_set(edata, edata_base_get(edata));
+	edata_zeroed_set(edata, false);
+
+	assert(edata_pai_get(edata) == EXTENT_PAI_HPA);
+	assert(edata_state_get(edata) == extent_state_active);
+	assert(edata_arena_ind_get(edata) == shard->ind);
+	assert(edata_szind_get_maybe_invalid(edata) == SC_NSIZES);
+	assert(!edata_slab_get(edata));
+	assert(edata_committed_get(edata));
+	assert(edata_base_get(edata) != NULL);
+
+	if (edata_ps_get(edata) != NULL) {
+		emap_deregister_boundary(tsdn, shard->hpa->central.emap, edata);
+
+		malloc_mutex_lock(tsdn, &shard->mtx);
+		edata_t *evicted_ps = psset_dalloc(&shard->psset, edata);
+		malloc_mutex_unlock(tsdn, &shard->mtx);
+
+		edata_cache_put(tsdn, shard->edata_cache, edata);
+
+		if (evicted_ps != NULL) {
+			bool err = emap_register_boundary(tsdn,
+			    shard->hpa->central.emap, evicted_ps, SC_NSIZES,
+			    /* slab */ false);
+			/*
+			 * Registration can only fail on OOM, but the boundary
+			 * mappings should have been initialized during
+			 * allocation.
+			 */
+			assert(!err);
+			edata_slab_set(evicted_ps, false);
+			edata_ps_set(evicted_ps, NULL);
+
+			assert(edata_arena_ind_get(evicted_ps) == shard->ind);
+			hpa_dalloc_central(tsdn, shard, evicted_ps);
+		}
+	} else {
+		hpa_dalloc_central(tsdn, shard, edata);
+	}
+}
+
+void
+hpa_shard_destroy(tsdn_t *tsdn, hpa_shard_t *shard) {
+	/*
+	 * By the time we're here, the arena code should have dalloc'd all the
+	 * active extents, which means we should have eventually evicted
+	 * everything from the psset, so it shouldn't be able to serve even a
+	 * 1-page allocation.
+	 */
+	if (config_debug) {
+		edata_t edata = {0};
+		malloc_mutex_lock(tsdn, &shard->mtx);
+		bool psset_empty = psset_alloc_reuse(&shard->psset, &edata,
+		    PAGE);
+		malloc_mutex_unlock(tsdn, &shard->mtx);
+		assert(psset_empty);
```

Comment:
nit: maybe also check some stats (if it's easy to do)?

Reply:
Done.